src/write_kosak.c

```diff
@@ -0,0 +1,2159 @@
+// characters printed. Remaining arguments are the same as snprintf. Does not
+// overwrite the bounds of the buffer under any circumstances. When successful,
+// leaves *buffer pointing directly at a terminating NUl character (just like
+// snprintf).
+//
+// This method is designed to allow the caller to do series of calls to
+// bufprintf, and only check for errors at the end of the series rather than
+// after every call. This leads to shorter, more readable code in exchange for
+// wasted CPU effort in the event of an error. Since errors are expected to be
+// rare, this is a worthwhile tradeoff.
+//
+// Two kinds of errors are possible, and are indicated by the value of *size.
+// 1. If the buffer fills up (because the number of characters fed to it either
+// reached or exceeded its capacity), *size will be 1. (This represents the
+// fact that there is just enough space left for the terminating NUL). Note that
+// a buffer which is exactly full is indistinguishable from a buffer that has
+// overflowed. This distinction does not matter for our purposes, and it is far
+// more convenient to treat the "buffer exactly full" case as though it was an
+// overflow rather than separating it out.
+//
+// 2. If vsprintf returns an error, *size will be forced to 0.
+//
+// Callers who do not care about this distinction can just check for *size > 1
+// as a success indication.
+//
+// Example usage:
+//   char buffer[1024];
+//   char *p = buffer;
+//   size_t s = sizeof(buffer);
+//   bufprintf(&p, &s, fmt, args...);
+//   bufprintf(&p, &s, fmt, args...);  /* add more */
+//   bufprintf(&p, &s, fmt, args...);  /* add yet more */
+//   /* check for errors here */
+//   if (s < 2) {
+//     ERROR("error (s==0) or overflow (s==1)");
+//     return -1;
+//   }
+static void bufprintf(char **buffer, size_t *size, const char *fmt, ...) {
+  if (*size == 0) {
+    return;
+  }
+  va_list ap;
+  va_start(ap, fmt);
+  int result = vsnprintf(*buffer, *size, fmt, ap);
+  va_end(ap);
+
+  if (result < 0) {
+    *size = 0;
+    return;
+  }
+  // If the result was *size or more, the output was truncated. In that case,
+  // adjust the pointer and size so they are pointing to the last byte (the
+  // terminating NUL).
+  if (result >= *size) {
+    result = *size - 1;
+  }
+  *buffer += result;
+  *size -= result;
+}
+
+//==============================================================================
+//==============================================================================
+//==============================================================================
+// Server info submodule.
+//==============================================================================
+//==============================================================================
+//==============================================================================
+typedef struct {
+  char project_id[128];
+  char instance_id[128];
+  char zone[128];
+} server_ctx_t;
```

Comment:
This is probably more correctly gce_instance_ctx_t.  We'll need something different as we start having this support AWS